### PR TITLE
Video thumbnails (optional)

### PR DESCRIPTION
**Pull request by benjaminhallock**
I copied the model from Photo to Video as best as I could. The new background image is optional as of now, the play icon is centered and scaled proportionally, and it shouldn't affect existing code. Added a thumbnail to the demo app. If I forgot to do additional testing, or screwed up this process, please help me out there.

### CODE REVIEW

**Comment by jessesquires** [comment]
Thanks @benjaminhallock ! I'll take a look soon :smile:

File: JSQMessagesViewController/Model/JSQVideoMediaItem.h [hunk 1]
```diff
@@ -32,12 +32,18 @@
 @property (nonatomic, strong) NSURL *fileURL;
 
 /**
+ * Adding a image to mask behind the video.
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.h]: let's say:

> The thumbnail image for the video, or `nil`.

File: JSQMessagesViewController/Model/JSQVideoMediaItem.h [hunk 2]
```diff
@@ -32,12 +32,18 @@
...
+ */
+@property (copy, nonatomic) UIImage *image;
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.h]: - let's make this `strong`, not copy.
- let's rename `thumbnailImage`

File: JSQMessagesViewController/Model/JSQVideoMediaItem.h [hunk 3]
```diff
@@ -53,4 +59,21 @@
...
+ *
+ *  @return An initialized `JSQVideoMediaItem` if successful, `nil` otherwise.
+ *
+ *  @discussion If the video must be downloaded from the network,
+ *  you may initialize a `JSQVideoMediaItem` with a `nil` fileURL or specify `NO` for
+ *  isReadyToPlay. Once the video has been saved to disk, or is ready to stream, you can
+ *  set the fileURL property or isReadyToPlay property, respectively. The background thumbnail 
+ *  is optional.
+ */
+- (instancetype)initWithFileURL:(NSURL *)fileURL isReadyToPlay:(BOOL)isReadyToPlay image:(UIImage *)image;
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.h]: rename last param to `thumbnailImage:`

File: JSQMessagesViewController/Model/JSQVideoMediaItem.h [hunk 4]
```diff
@@ -53,4 +59,21 @@
...
+ *  @param image         The background thumbnail for the video.
+ *
+ *  @return An initialized `JSQVideoMediaItem` if successful, `nil` otherwise.
+ *
+ *  @discussion If the video must be downloaded from the network,
+ *  you may initialize a `JSQVideoMediaItem` with a `nil` fileURL or specify `NO` for
+ *  isReadyToPlay. Once the video has been saved to disk, or is ready to stream, you can
+ *  set the fileURL property or isReadyToPlay property, respectively. The background thumbnail 
+ *  is optional.
+ */
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.h]: :+1:

File: JSQMessagesViewController/Model/JSQVideoMediaItem.m [hunk 5]
```diff
@@ -35,6 +35,18 @@ @implementation JSQVideoMediaItem
...
+        _isReadyToPlay = isReadyToPlay;
+        _image = [image copy];
+        _cachedVideoImageView = nil;
+    }
+    return self;
+}
+
 - (instancetype)initWithFileURL:(NSURL *)fileURL isReadyToPlay:(BOOL)isReadyToPlay
 {
     self = [super init];
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.m]: let's have this `init` call your new one with a `nil` image :smile:

File: JSQMessagesViewController/Model/JSQVideoMediaItem.m [hunk 6]
```diff
@@ -79,20 +97,51 @@ - (UIView *)mediaView
...
-        UIImageView *imageView = [[UIImageView alloc] initWithImage:playIcon];
-        imageView.backgroundColor = [UIColor blackColor];
-        imageView.frame = CGRectMake(0.0f, 0.0f, size.width, size.height);
-        imageView.contentMode = UIViewContentModeCenter;
-        imageView.clipsToBounds = YES;
-        [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:imageView isOutgoing:self.appliesMediaViewMaskAsOutgoing];
-        self.cachedVideoImageView = imageView;
+
+    if (self.cachedVideoImageView == nil)
+    {
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.m]: `{` curly braces on same line :smile:

File: JSQMessagesViewController/Model/JSQVideoMediaItem.m [hunk 7]
```diff
@@ -79,20 +97,51 @@ - (UIView *)mediaView
...
+            CGSize size = [self mediaViewDisplaySize];
+            UIImage *playIcon = [[UIImage jsq_defaultPlayImage] jsq_imageMaskedWithColor:[UIColor lightGrayColor]];
+
+            UIImageView *imageView = [[UIImageView alloc] initWithImage:playIcon];
+            imageView.backgroundColor = [UIColor blackColor];
+            imageView.frame = CGRectMake(0.0f, 0.0f, size.width, size.height);
+            imageView.contentMode = UIViewContentModeCenter;
+            imageView.clipsToBounds = YES;
+            [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:imageView isOutgoing:self.appliesMediaViewMaskAsOutgoing];
+            self.cachedVideoImageView = imageView;
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.m]: let's pull this out into a new private helper method:

`configureVideoViewWithDefaultPlayIcon`

File: JSQMessagesViewController/Model/JSQVideoMediaItem.m [hunk 8]
```diff
@@ -79,20 +97,51 @@ - (UIView *)mediaView
...
+            [playIcon drawInRect:scaledImageRect];
+            UIImage *resultImage = UIGraphicsGetImageFromCurrentImageContext();
+            UIGraphicsEndImageContext();
+
+            UIImageView *imageView = [[UIImageView alloc] initWithImage:resultImage];
+            imageView.frame = CGRectMake(0.0f, 0.0f, size.width, size.height);
+            imageView.contentMode = UIViewContentModeScaleAspectFill;
+            imageView.clipsToBounds = YES;
+            [JSQMessagesMediaViewBubbleImageMasker applyBubbleImageMaskToMediaView:imageView isOutgoing:self.appliesMediaViewMaskAsOutgoing];
+            self.cachedVideoImageView = imageView;
```
Comment (jessesquires) [inline JSQMessagesViewController/Model/JSQVideoMediaItem.m]: another helper method for the `else` block, too:

`configureVideoViewWithCustomThumbnail`

**Comment by jessesquires** [comment]
Thanks again @benjaminhallock ! :+1: 

Left a few comments to address, then we can merge!

**Comment by benjaminhallock** [comment]
Oh boy, this is over my head but I'll see if I can fix this and merge.

**Comment by jessesquires** [comment]
@benjaminhallock no worries! :smile: 

You should be able to:

1. make changes
2. commit
3. push

**Comment by jessesquires** [comment]
Thanks @benjaminhallock !

We included your changes in #1823

**Pull request closed without merging**